### PR TITLE
add exploit prediction scoring system support

### LINKED ISSUE #193: Integrate Exploit Prediction Scoring System to Enrich CVE Data
The purpose of this issue is to propose an enhancement for the CveXplore project. The enhancement involves the integration of an exploit prediction scoring system, which would serve to enrich the CVE data with predictive insights regarding the potential for exploitation.

### Advantages:

* Enhanced risk assessment capabilities for end-users, providing a clearer understanding of the exploitability of vulnerabilities.
* Improved prioritization of vulnerabilities to patch, based on a quantifiable exploitation likelihood score.

### PR DESCRIPTION
**Pull request by acuas**
fixes #193

### CODE REVIEW

File: CveXplore/common/config.py [hunk 1]
```diff
@@ -78,6 +78,8 @@ class Configuration(object):
         "cwe": "https://cwe.mitre.org/data/xml/cwec_latest.xml.zip",
         "capec": "https://capec.mitre.org/data/xml/capec_latest.xml",
         "via4": "https://www.cve-search.org/feeds/via4.json",
+        # See EPSS at https://www.first.org/epss
+        "epss": "https://epss.cyentia.com/epss_scores-current.csv.gz"
```
Comment (P-T-I) [inline CveXplore/common/config.py:82]: Who is hosting this?

File: CveXplore/database/maintenance/Sources_process.py [hunk 2]
```diff
@@ -1094,6 +1094,80 @@ def populate(self, **kwargs):
...
+    @staticmethod
+    def process_epss_item(item=None):
+        if item is None:
+            return None
+
+        epss = {
+            "id": item[0],
+            "epss": item[1],
+            "percentile": item[2]
+        }
```
Comment (P-T-I) [inline CveXplore/database/maintenance/Sources_process.py:1116]: Looking at the data that is imported; why would you go for a separate collection? Looking at the rest of the code it looks like there is a one-to-one match between the CVE's and the EPSS, wouldn't it make sense to store the EPSS information straight into the cves collection?

File: CveXplore/common/config.py (hunk 1 above)
Comment (acuas) [reply CveXplore/common/config.py:82]: https://www.cyentia.com/about/#who-we-are -> It appears to be a cybersecurity research company, but I found the EPSS project on the first.org website, and it mentioned that:

"EPSS is an emerging standard developed by a volunteer group of researchers, practitioners, academics and government personnel. We grant the use of EPSS scores freely to the public, subject to the conditions below. We reserve the right to update the model and these webpages periodically, as necessary, though we will make every attempt to provide sufficient notice to users in the event of material changes. While membership in the EPSS SIG is not required to use or implement EPSS, however, we ask that if you are using EPSS, that you provide appropriate attribution where possible. EPSS can be cited either from this website (e.g. "See EPSS at https://www.first.org/epss), or as: Jay Jacobs, Sasha Romanosky, Benjamin Edwards, Michael Roytman, Idris Adjerid, (2021), Exploit Prediction Scoring System, Digital Threats Research and Practice, 2(3)"

So, if you are concerned about licensing, it should not be a problem.

File: CveXplore/database/maintenance/Sources_process.py (hunk 2 above)
Comment (acuas) [reply CveXplore/database/maintenance/Sources_process.py:1116]: You are correct. When I examined the code, I realized that I could implement it similarly to the via4 implementation. I didn't consider checking the data from via4, and that's my mistake. I will make the necessary changes to store the EPSS information directly into the CVEs collection.

File: CveXplore/common/config.py (hunk 1 above)
Comment (P-T-I) [reply CveXplore/common/config.py:82]: Licensing is not my main concern; I'm more concerned about the site where the epss is downloaded from; isn't there a download available straight from the first.org site?

Comment (P-T-I) [reply CveXplore/common/config.py:82]: I thought this link was an "external" link, but I just discovered that the link is posted on the first.org site.

File: CveXplore/database/maintenance/Sources_process.py [hunk 3]
```diff
@@ -1094,6 +1094,80 @@ def populate(self, **kwargs):
...
+        self.feed_url = Configuration.getFeedURL(self.feed_type.lower())
+        self.logger = logging.getLogger(__name__)
+        self.is_update = True
+
+    @staticmethod
+    def process_epss_item(item=None):
+        if item is None:
+            return None
+
+        epss = {
```
Comment (P-T-I) [inline CveXplore/database/maintenance/Sources_process.py:1112]: Additional comment; might be useful to add the epss date as well? They release on a daily schedule, right?

Comment (acuas) [reply CveXplore/database/maintenance/Sources_process.py:1112]: Yes, they release it on a daily schedule. It sounds good, I will add the date as well.

Comment (P-T-I) [reply CveXplore/database/maintenance/Sources_process.py:1112]: Awesome!

**Comment by acuas** [comment]
@P-T-I  I updated the PR, I will let some comments on it on which parts I am not sure about the implementation.

File: CveXplore/database/maintenance/Sources_process.py [hunk 4]
```diff
@@ -1024,6 +1024,71 @@ def populate(self, **kwargs):
...
+                "percentile": item[2],
+                "lastModified": date.strftime('%Y-%m-%dT%H:%M:%SZ')
+            }
+        }
+
+        return epss
+
+    def process_item(self, item):
+        # overwrite the feed_type, so that the correct collection is used
+        self.feed_type = self.cve_collection
```
Comment (acuas) [inline CveXplore/database/maintenance/Sources_process.py:1056]: I don't know if I can do something better than this. I don't like the fact that I overwrite the feed_type, after the  download phase is done.

Comment (P-T-I) [reply CveXplore/database/maintenance/Sources_process.py:1056]: Testing this PR at the moment; will perform some changes (if any) in the PR. Looking at this particular comment I am not convinced that overwriting the self.feed_type has anything to do with the items getting into the correct collection. I believe the feed_type is used when writing the lastModified values into the database.

Comment (acuas) [reply CveXplore/database/maintenance/Sources_process.py:1056]: I don't think so, if you look at the insertion function from the DownloadHelper class -> https://github.com/cve-search/CveXplore/blob/master/CveXplore/database/maintenance/DownloadHandler.py#L157-L172 . You can see that self.feed_type is used for insertion into mongo. Specifying the collection in DatabaseAction is useless I think.

Comment (P-T-I) [reply CveXplore/database/maintenance/Sources_process.py:1056]: You are correct; the collection specification in DatabaseAction is not used; I'll have to have a closer look!

**Comment by P-T-I** [comment]
@acuas Overall the PR looked ok; created a separate branch with the changes I've got implemented; https://github.com/cve-search/CveXplore/tree/epss_pr_draft. Thought it was easier to diff those and merge into you branch.

**Comment by acuas** [comment]
Thank you! I will merge it into our branch.

**Comment by acuas** [comment]
@P-T-I rebased into our branch.

**Review by P-T-I — APPROVED** [review]

**Comment by P-T-I** [comment]
@acuas Thank you very much! It's a really nice addition!

**Comment by acuas** [comment]
@P-T-I Thank you as well for your time and dedication!

**Pull request merged**